syncer/syncer.go

```diff
@@ -0,0 +1,402 @@
+	vDone := make(chan struct{})
+	go s.startValidating(ctx, s.run, vQueue, vDone)
+	defer func() {
+		close(vQueue)
+		<-vDone
+		logger.With().Info("node is synced. validation done",
+			log.FieldNamed("current", s.ticker.GetCurrentLayer()),
+			log.FieldNamed("latest", s.mesh.LatestLayer()),
+			log.FieldNamed("validated", s.mesh.ProcessedLayer()))
+		logger.Info("finished sync run #%v", s.run)
+		s.endSyncProcess()
+	}()
+
+	// using ProcessedLayer() instead of LatestLayer() so we can validate layers on a best-efforts basis.
+	// TODO: is it safe to skip layer 0
+	for layerID := s.mesh.ProcessedLayer() + 1; layerID <= s.ticker.GetCurrentLayer(); layerID++ {
+		var layer *types.Layer
+		var err error
+		if layer, err = s.syncLayer(ctx, layerID); err != nil {
+			logger.With().Error("failed to sync to layer", log.Err(err))
+			return false
+		}
+
+		if len(layer.Blocks()) == 0 {
+			logger.Info("setting layer %v to zero-block", layerID)
+			if err := s.mesh.SetZeroBlockLayer(layerID); err != nil {
+				logger.With().Error("failed to set zero-block for layer", layerID, log.Err(err))
+				return false
+			}
+		}
+
+		if s.shouldValidateLayer(layerID) {
+			vQueue <- layer
+		}
+		logger.Info("finished data sync for layer %v", layerID)
+	}
+	s.setStateAfterSync(ctx)
+	logger.With().Info("node is synced. waiting for validation",
+		log.FieldNamed("current", s.ticker.GetCurrentLayer()),
+		log.FieldNamed("latest", s.mesh.LatestLayer()),
+		log.FieldNamed("validated", s.mesh.ProcessedLayer()))
+	return true
+}
+
+func (s *Syncer) setStateBeforeSync(ctx context.Context) {
+	current := s.ticker.GetCurrentLayer()
+	if current == 0 {
+		s.setSyncState(ctx, synced)
+	}
+	latest := s.mesh.LatestLayer()
+	if current > latest && current-latest >= outOfSyncThreshold {
+		s.logger.WithContext(ctx).With().Info("node is too far behind",
+			log.FieldNamed("current", current),
+			log.FieldNamed("latest", latest),
+			log.FieldNamed("behindThreshold", outOfSyncThreshold))
+		s.setSyncState(ctx, notSynced)
+	}
+}
+
+func (s *Syncer) setStateAfterSync(ctx context.Context) {
+	currSyncState := s.getSyncState()
+	current := s.ticker.GetCurrentLayer()
+	// if we have gossip-synced to the target synced layer, we are ready to participate in the consensus
+	if currSyncState == gossipSync && s.getTargetSyncedLayer() <= current {
+		s.setSyncState(ctx, synced)
+	} else if currSyncState == notSynced {
+		// wait till s.ticker.GetCurrentLayer() + numGossipSyncLayers to participate in consensus
+		s.setSyncState(ctx, gossipSync)
+		s.setTargetSyncedLayer(ctx, current+numGossipSyncLayers)
+	}
+}
+
+func (s *Syncer) syncLayer(ctx context.Context, layerID types.LayerID) (*types.Layer, error) {
+	if s.isClosed() {
+		return nil, errors.New("shutdown")
+	}
+
+	layer, err := s.getLayerFromPeers(ctx, layerID)
+	if err != nil {
+		s.logger.WithContext(ctx).With().Error("failed to get layer from neighbors", layerID, log.Err(err))
```

Comment:
do we have a test for this case? failure to sync

Reply:
yes. i renamed it to `TestSynchronize_getLayerFromPeersFailed` to be clear.